File: lib/mqsco.js
"use strict";
/*
  Copyright (c) IBM Corporation 2017, 2023

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.

   Contributors:
     Mark Taylor - Initial Contribution
*/

/*
 * MQSCO is a JavaScript object containing the fields we need for the MQSCO
 * in a more idiomatic style than the C definition - in particular for
 * fixed length character buffers.
 */


// Import MQI definitions
const MQC        = require("./mqidefs.js");

/*
 * This constructor sets all the common fields for the structure.
 */
/**
 * This constructor sets default values for the object.
 * @class
 * @classdesc
 * This is a class containing the fields needed for the MQSCO
 * (MQ SSL/TLS Configuration Options) structure. See the
 * {@link https://www.ibm.com/docs/en/ibm-mq/latest?topic=mqi-mqsco-ssltls-configuration-options|MQ Documentation}
 * for more details on the usage of each field.
 */
exports.MQSCO  = function () {
  /** @member {String} */
  this.KeyRepository          = null;
  /** @member {String} */
  this.CryptoHardware         = null;
  /** @member {number} */
  this.KeyResetCount          = MQC.MQSCO_RESET_COUNT_DEFAULT;
  /** @member {boolean} */
  this.FipsRequired           = false;
  /** @member {Array} */
  this.EncryptionPolicySuiteB = [
    MQC.MQ_SUITE_B_NONE,          MQC.MQ_SUITE_B_NOT_AVAILABLE,
    MQC.MQ_SUITE_B_NOT_AVAILABLE, MQC.MQ_SUITE_B_NOT_AVAILABLE];
  /** @member {number} */
  this.CertificateValPolicy   = MQC.MQ_CERT_VAL_POLICY_DEFAULT;
  /** @member {String} */
  this.CertificateLabel       = null;
  /** @member {String} */
  this.KeyRepoPassword        = null;
  /** @member {String} */
  this.HTTPSKeyStore          = null;
  /** @member {number} */
  this.HTTPSCertValidation    = 0;
  /** @member {number} */
  this.HTTPSCertRevocation    = 0;
  Object.seal(this);
};

File: src/mqsco.cc
/*
  Copyright (c) IBM Corporation 2017, 2024

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.

   Contributors:
     Mark Taylor - Initial Contribution
*/

#include "mqi.h"

/* Conversion routines for the MQSCO structure */

void copySCOtoC(Env env, Object jssco, PMQSCO pmqsco) {

  pmqsco->Version = MQSCO_VERSION_4; // Assume this base level

  setMQIString(env,pmqsco->KeyRepository, jssco, "KeyRepository",MQ_SSL_KEY_REPOSITORY_LENGTH);
  setMQIString(env,pmqsco->CryptoHardware,jssco, "CryptoHardware", MQ_SSL_CRYPTO_HARDWARE_LENGTH);
  pmqsco->KeyResetCount     = getMQLong(jssco,"KeyResetCount");
  pmqsco->FipsRequired      = jssco.Get("FipsRequired").As<Boolean>()?MQSSL_FIPS_YES:MQSSL_FIPS_NO;

  Array a = jssco.Get("EncryptionPolicySuiteB").As<Array>();
  for (int i = 0;i<4;i++) {
    Value v = a[i];
    pmqsco->EncryptionPolicySuiteB[i] = v.As<Number>().Int32Value();
  }
  pmqsco->CertificateValPolicy = getMQLong(jssco,"CertificateValPolicy");
  setMQIString(env, pmqsco->CertificateLabel , jssco,"CertificateLabel",MQ_CERT_LABEL_LENGTH);
  if (pmqsco->CertificateLabel[0] != ' ' && pmqsco->CertificateLabel[0] != 0) {
    if (pmqsco->Version < MQSCO_VERSION_5) {
      pmqsco->Version = MQSCO_VERSION_5;
    }
  }

  Value v = jssco.Get("KeyRepoPassword");
  if (v.IsString()) {
    pmqsco->KeyRepoPasswordPtr = mqnStrdup(env,jssco.Get("KeyRepoPassword").As<String>().Utf8Value().c_str());
    pmqsco->KeyRepoPasswordOffset = 0;
    pmqsco->KeyRepoPasswordLength = strlen((char *)pmqsco->KeyRepoPasswordPtr);
    if (pmqsco->Version < MQSCO_VERSION_6) {
      pmqsco->Version = MQSCO_VERSION_6;
    }
  }

  v = jssco.Get("HTTPSKeyStore");
  if (v.IsString()) {
    pmqsco->HTTPSKeyStorePtr = mqnStrdup(env,jssco.Get("HTTPSKeyStore").As<String>().Utf8Value().c_str());
    pmqsco->KeyRepoPasswordOffset = 0;
    pmqsco->KeyRepoPasswordLength = strlen((char *)pmqsco->HTTPSKeyStorePtr);
    if (pmqsco->Version < MQSCO_VERSION_7) {
      pmqsco->Version = MQSCO_VERSION_7;
    }
  }

  pmqsco->HTTPSCertValidation = getMQLong(jssco,"HTTPSCertValidation");
  pmqsco->HTTPSCertRevocation = getMQLong(jssco,"HTTPSCertRevocation");
  if (pmqsco->HTTPSCertRevocation != 0 || pmqsco->HTTPSCertValidation != 0) {
    if (pmqsco->Version < MQSCO_VERSION_7) {
      pmqsco->Version = MQSCO_VERSION_7;
    }
  }
  return;
};

void copySCOfromC(Env env, Object jssco, PMQSCO pmqsco) {
  return;
};

void cleanupSCO(PMQSCO pmqsco) {
  if (pmqsco) {
    mqnFreeString(pmqsco->KeyRepoPasswordPtr);
    mqnFreeString(pmqsco->HTTPSKeyStorePtr);
  }
  return;
}
